### PR TITLE
Flatness checking

### LINKED ISSUE #416: Method for checking flatness
For Z3-Noodler we need to check if an NFA is flat. It could be useful to add something like `is_flat` method for checking flatness of the automaton.

### PR DESCRIPTION
**Pull request by vhavlena**
Adding a new method for checking flatness of an NFA (`is_flat`).

### CODE REVIEW

File: include/mata/nfa/nfa.hh [hunk 1]
```diff
@@ -320,6 +320,15 @@ public:
      */
     bool is_acyclic() const;
 
+    /**
+     * @brief Is the automaton flat?
+     * Flat automaton is an NFA whose every SCC is a simple loop. Basically each state in an
```
Comment (Adda0) [inline include/mata/nfa/nfa.hh]: A `@brief` section must be ended with a empty line (containing only `     *`), or a start of another section. Therefore, we need to add a newline between these two lines.

File: src/nfa/nfa.cc [hunk 2]
```diff
@@ -379,6 +379,34 @@ bool Nfa::is_acyclic() const {
...
+
+    mata::nfa::Nfa::TarjanDiscoverCallback callback {};
+    callback.scc_discover = [&](const std::vector<mata::nfa::State>& scc, const std::vector<mata::nfa::State>& tarjan_stack) -> bool {
+        (void)tarjan_stack;
+        
+        for(const mata::nfa::State& st : scc) {
+            bool one_input_visited = false;
+            for (const mata::nfa::SymbolPost& sp : this->delta[st]) {
+                for (const mata::nfa::State& tgt : scc) {
+                    if(sp.targets.find(tgt) != sp.targets.end()) {
```
Comment (Adda0) [inline src/nfa/nfa.cc]:
```suggestion
                    if(sp.targets.contains(tgt)) {
```
The condition can be simplified.

**Review by Adda0 — APPROVED** [review]
Looks great. I like that the implementation of the Tarjan's algorithm allows for these operations thanks to the callbacks.

File: include/mata/nfa/nfa.hh (hunk 1 above)
Comment (vhavlena) [reply include/mata/nfa/nfa.hh]: Ok, fixed

**Review by jurajsic — APPROVED** [review]

**Pull request merged**